packages/client/src/queries/applyQuery.ts

```diff
@@ -0,0 +1,331 @@
+      return undefined as unknown as QueryReturnType<typeof responseDataType>;
+    } else {
+      throw new Error("Got null response when nullable was not allowed");
+    }
+  }
+
+  if (
+    responseDataType.multiplicity != null
+    && responseDataType.multiplicity !== false
+  ) {
+    const withoutMultiplicity = { ...responseDataType, multiplicity: false };
+    for (let i = 0; i < responseValue.length; i++) {
+      responseValue[i] = remapQueryResponse(
+        responseValue[i],
+        withoutMultiplicity,
+        client,
+        definitions,
+      );
+    }
+    return responseValue as QueryReturnType<typeof responseDataType>;
+  }
+
+  switch (responseDataType.type) {
+    case "union": {
+      throw new Error("Union return types are not yet supported");
+    }
+
+    case "set": {
+      for (let i = 0; i < responseValue.length; i++) {
+        responseValue[i] = remapQueryResponse(
+          responseValue[i],
+          responseDataType.set,
+          client,
+          definitions,
+        );
+      }
+
+      return responseValue as QueryReturnType<typeof responseDataType>;
+    }
+
+    case "attachment": {
+      return createAttachmentFromRid(client, responseValue) as QueryReturnType<
+        typeof responseDataType
+      >;
+    }
+    case "object": {
+      const def = definitions.get(responseDataType.object);
+      if (!def) {
+        throw new Error(
+          `Missing definition for ${responseDataType.object}`,
+        );
+      }
+      return createQueryObjectResponse(
+        responseValue,
+        def,
+      ) as QueryReturnType<
+        typeof responseDataType
+      >;
+    }
+
+    case "objectSet": {
+      const def = definitions.get(responseDataType.objectSet);
+      if (!def) {
+        throw new Error(
+          `Missing definition for ${responseDataType.objectSet}`,
+        );
+      }
+      if (typeof responseValue === "string") {
+        return createObjectSet(def, client, {
+          type: "intersect",
+          objectSets: [
+            { type: "base", objectType: responseDataType.objectSet },
+            { type: "reference", reference: responseValue },
+          ],
+        }) as QueryReturnType<typeof responseDataType>;
+      }
+
+      return createObjectSet(
+        def,
+        client,
+        responseValue,
+      ) as QueryReturnType<
+        typeof responseDataType
+      >;
+    }
+    case "struct": {
+      // figure out what keys need to be fixed up
+      for (const [key, subtype] of Object.entries(responseDataType.struct)) {
+        if (requiresConversion(subtype)) {
+          responseValue[key] = remapQueryResponse(
+            responseValue[key],
+            subtype,
+            client,
+            definitions,
+          );
+        }
+      }
+
+      return responseValue as QueryReturnType<typeof responseDataType>;
+    }
+    case "twoDimensionalAggregation": {
+      const result: { key: any; value: any }[] = [];
```

Comment:
These are any's right now, once I clean up the 2d/3d aggs type can clean this up